File: src/integration/jackson/src/jvmMain/java/community/flock/wirespec/integration/jackson/v3/java/WirespecSerialization.java
package community.flock.wirespec.integration.jackson.v3.java;

import community.flock.wirespec.java.Wirespec.Serialization;
import community.flock.wirespec.java.serde.DefaultParamSerialization;
import community.flock.wirespec.java.serde.DefaultPathSerialization;
import tools.jackson.databind.JavaType;
import tools.jackson.databind.json.JsonMapper;

import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;

/**
 * A reusable implementation of Wirespec.Serialization backed by Jackson 3.
 *
 * It rebuilds the supplied (immutable) {@link JsonMapper} with the
 * {@link WirespecModuleJava} registered. Java records expose their components through
 * accessors, so no visibility configuration is required. Parameter and path
 * serialization are delegated to the default Wirespec serializers.
 */
public class WirespecSerialization implements Serialization, DefaultParamSerialization, DefaultPathSerialization {

    private final JsonMapper wirespecObjectMapper;

    public WirespecSerialization(JsonMapper jsonMapper) {
        // The reserved-keyword naming strategy must be set on the builder
        // (Jackson 3 modules cannot register one).
        this.wirespecObjectMapper = jsonMapper.rebuild()
                .propertyNamingStrategy(new WirespecModuleJava.JavaReservedKeywordNamingStrategy())
                .addModule(new WirespecModuleJava())
                .build();
    }

    @Override
    public <T> byte[] serializeBody(T body, Type type) {
        if (body instanceof String) {
            return ((String) body).getBytes(StandardCharsets.UTF_8);
        } else {
            return wirespecObjectMapper.writeValueAsBytes(body);
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public <T> T deserializeBody(byte[] raw, Type valueType) {
        if (raw == null) {
            return null;
        }

        if (valueType.equals(String.class)) {
            return (T) new String(raw, StandardCharsets.UTF_8);
        } else {
            JavaType type = wirespecObjectMapper.constructType(valueType);
            return wirespecObjectMapper.readValue(raw, type);
        }
    }
}

File: src/integration/jackson/src/jvmMain/kotlin/community/flock/wirespec/integration/jackson/v3/kotlin/WirespecSerialization.kt
package community.flock.wirespec.integration.jackson.v3.kotlin

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.PropertyAccessor
import community.flock.wirespec.kotlin.Wirespec
import community.flock.wirespec.kotlin.Wirespec.Serialization
import community.flock.wirespec.kotlin.serde.DefaultParamSerialization
import community.flock.wirespec.kotlin.serde.DefaultPathSerialization
import tools.jackson.databind.json.JsonMapper
import kotlin.reflect.KType
import kotlin.reflect.javaType

/**
 * A reusable implementation of [Wirespec.Serialization] backed by Jackson 3.
 *
 * It rebuilds the supplied (immutable) [JsonMapper] with the [WirespecModuleKotlin]
 * registered and field-level visibility configured, so Wirespec models serialize by
 * field exactly like the Jackson 2 integration. Parameter and path serialization are
 * delegated to the default Wirespec serializers.
 */
class WirespecSerialization(
    jsonMapper: JsonMapper,
) : Serialization,
    Wirespec.ParamSerialization by DefaultParamSerialization(),
    Wirespec.PathSerialization by DefaultPathSerialization() {

    private val wirespecObjectMapper: JsonMapper = jsonMapper.rebuild()
        .changeDefaultVisibility {
            it
                .withVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE)
                .withFieldVisibility(JsonAutoDetect.Visibility.ANY)
        }
        .propertyNamingStrategy(KotlinReservedKeywordNamingStrategy())
        .addModule(WirespecModuleKotlin())
        .build()

    override fun <T : Any> serializeBody(t: T, kType: KType): ByteArray = when (t) {
        is String -> t.toByteArray()
        else -> wirespecObjectMapper.writeValueAsBytes(t)
    }

    @Suppress("UNCHECKED_CAST")
    @OptIn(ExperimentalStdlibApi::class)
    override fun <T : Any> deserializeBody(raw: ByteArray, kType: KType): T = when {
        kType.classifier == String::class -> raw as T
        else ->
            wirespecObjectMapper
                .constructType(kType.javaType)
                .let { wirespecObjectMapper.readValue(raw, it) }
    }
}
